primitives/trie/src/storage_proof.rs

```diff
@@ -14,65 +14,990 @@
+					=> StorageProof::KnownQueryPlanAndValues(Decode::decode(value)?),
+				StorageProofKind::Full => StorageProof::Full(Decode::decode(value)?),
+				StorageProofKind::TrieSkipHashesForMerge
+					=> return Err(codec::Error::from("Invalid storage kind")),
+				StorageProofKind::TrieSkipHashesFull
+					=> StorageProof::TrieSkipHashesFull(Decode::decode(value)?),
+		})
+	}
+}
+
+impl Encode for StorageProof {
+	fn encode_to<T: CodecOutput>(&self, dest: &mut T) {
+		(self.kind() as u8).encode_to(dest);
+		match self {
+			StorageProof::Flatten(p) => p.encode_to(dest),
+			StorageProof::TrieSkipHashes(p) => p.encode_to(dest),
+			StorageProof::KnownQueryPlanAndValues(p) => p.encode_to(dest),
+			StorageProof::Full(p) => p.encode_to(dest),
+			StorageProof::TrieSkipHashesFull(p) => p.encode_to(dest),
+			StorageProof::TrieSkipHashesForMerge(..) => panic!("merge did not recurse as told"),
+		}
 	}
+}
+
+/// This encodes the full proof capabillity under
+/// legacy proof format.
+pub struct LegacyEncodeAdapter<'a>(pub &'a StorageProof);
 
+impl<'a> Encode for LegacyEncodeAdapter<'a> {
+	fn encode_to<T: CodecOutput>(&self, dest: &mut T) {
+		0u8.encode_to(dest);
+		self.0.encode_to(dest);
+	}
+}
+
+/// This encodes only if storage proof is a flatten proof.
+/// It panics otherwhise, so it should only be use when we
+/// got strong guaranties of the proof kind.
+pub struct FlattenEncodeAdapter<'a>(pub &'a StorageProof);
+
+impl<'a> Encode for FlattenEncodeAdapter<'a> {
+	fn encode_to<T: CodecOutput>(&self, dest: &mut T) {
+		match self.0 {
+			StorageProof::Flatten(nodes) => nodes.encode_to(dest),
+			_ => panic!("Usage of flatten encoder on non flatten proof"),
+		}
+	}
+}
+
+#[cfg_attr(test, derive(Debug, PartialEq, Eq))]
+/// Decode variant of `LegacyEncodeAdapter`.
+pub struct LegacyDecodeAdapter(pub StorageProof);
+
+/// Allow read ahead on input by chaining back some already consumed data.
+pub struct InputRevertPeek<'a, I>(pub &'a mut &'a [u8], pub &'a mut I);
+
+impl<'a, I: CodecInput> CodecInput for InputRevertPeek<'a, I> {
+	fn remaining_len(&mut self) -> CodecResult<Option<usize>> {
+		Ok(self.1.remaining_len()?.map(|l| l + self.0.len()))
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> CodecResult<()> {
+		let mut offset = 0;
+		if self.0.len() > 0 {
+			if self.0.len() > into.len() {
+				into.copy_from_slice(&self.0[..into.len()]);
+				*self.0 = &self.0[into.len()..];
+				return Ok(());
+			} else {
+				into[..self.0.len()].copy_from_slice(&self.0[..]);
+				*self.0 = &[][..];
+				offset = self.0.len();
+			}
+		}
+		self.1.read(&mut into[offset..])
+	}
+
+	fn read_byte(&mut self) -> CodecResult<u8> {
+		if self.0.len() > 0 {
+			let result = self.0[0];
+			*self.0 = &self.0[1..];
+			Ok(result)
+		} else {
+			self.1.read_byte()
+		}
+	}
+}
+
+impl Decode for LegacyDecodeAdapter {
+	fn decode<I: CodecInput>(value: &mut I) -> CodecResult<Self> {
+		let legacy = value.read_byte()?;
+		Ok(if legacy == 0 {
```

Comment:
proof starting with 0 is the empty flatten proof if there is only one byte of content or the proof encoded afterward if there is more content, but will probably remove this encoding as it was here to facilitate transition, but even in the description of the PR I really started to question if it could be of any use.